README.md
# schapiuino

## Overview

A basic library to handle a Client HTTP using an Arduino UNO and an Ethernet Shield.
It wrappes Ethernet setup, begin, connect, and send/receive request/response.

## Contents

- ClientHandler.h: the Client Handler Class definition file to create HTTP requests.
- ClientHandler.cpp: Client Handler implementation file
- examples:
	- ClientHandler_Basic_GET: a basic example making a GET to a defined serverUrl, path, using header Accept: text/plain.
	- ClientHandler_LCD: a basic example making a GET to server and showing response on LCD.
- ServerHandler.h: the Server Handler Class definition file to listen HTTP request and create responses.
- ServerHandler.cpp: Server Handler implementation file
- examples:
    - Server_Basic: Basic example that responses avec a basic HTML file on GET and POST method
    - Server_Servo: Example exposing an API to manage a Servo Motor.


## API

[Client Handler](#clientHandler)

[Server Handler](#serverHandler)

## <a name="clientHandler"></a>Client Handler

### ClientHandler(EthernetClient *client, byte mac[], int ipRanges[]) - Constructor
To init a ClientHandler object, you need to specify
* **client**: Pointer to an EthernetClient
* **mac**: Mac address to be used by the EthernetShield
* **ip**: Array containing the ranges to set the IP for the EthernetShield; Example [192, 0, 0, 1]

### bool  init()
To be used within the setup function. It initialises the EthernetShield with the values passed in the constructor

### bool    initClientHandler(String httpMethod, char *server, int severPort, String path)
Set up the parameters to make an HTTP request. Currently it doesn't accept body for the request. It returns true if the client could be connected to the server. 

* **httpMethod**: Http method to be used in the request: GET, POST, PUT, DELETE, PATCH
* **server**: serverUrl or IP
* **serverPort**: serverPort, normally 80
* **path**: Path to the requested service

Attention: it is not because the request was initialise that it was already send. This method initialises the [request String](http://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html) but it doesn't set up the end of it. **See send method**.

### void    addHeader(String header)
Allow addition of headers to make the request. The string cotains the key:value format. Example: "Accept:application/json"

### const char* send()
Sends the request to the server by adding "Connection: close" header and a CLRF to the request String. It returns the body catched in the server response in a const char* format.

### int     getResponseStatusCode()
Returns the Response's status code

### const char* getResponseBody()
Return the Response's body

### void    enableLogs()
Allow inner logs to be show in the serial monitor

###void    disableLogs()
Disable inner logs to be show in the serial monitor. This is the default behavior

## <a name="serverHandler"></a>Server Handler

### ServerHandler(EthernetServer *server, byte mac[], int ipRanges[]) - Constructor
To init a ClientHandler object, you need to specify:
* **server**: Pointer to an EthernetServer object
* **mac**: Mac address to be used by the EthernetShield
* **ip**: Array containing the ranges to set the IP for the EthernetShield; Example [192, 0, 0, 1]

### void    init()
To be used within the setup function. It initialises the EthernetShield with the values passed in the constructor.

### bool    listenRequest()
Listen if a request is incoming. If it is, it blocks the process and start reading the request.
When it returns true, the request is already read and ready to be analysed.

### String      getHttpMethod()
Returns the HTTP Method used by the incoming request.

### String      getRequestPath()
Returns the path used by the incoming request.

### const char* getRequestHeaders()
Returns all the headers given by the incoming request.

### const char* getRequestBody()
Returns the body (payload) sent in the incoming request.

### ServerHandler & buildResponse(int statusCode, String message = "")
Set the beginning of the response line, setting status code.
* **status**: HTTP status code to send in the response.
* **message**: (OPTIONAL): You can use another code (not in the http status codes list, in this case, set a message to explain it)

### ServerHandler & appendHeaderResponse(String header)
Add a header to the response.

### ServerHandler & appendBodyResponse(String body)
Add a body to the response if wanted. It should be called once!. If it is used, don't forget to add the "Content-Type" header before.

### void    send()
It ends the request line, close the connection and send the client the response.

*buildResponse, appendHeaderResponse and appendBodyResponse are **chainable** methods, so you can write your code like this :*

```
responseHandler
    .buildResponse(201)
    .appendHeaderResponse("Content-Type: application/json")
    .appendHeaderResponse("Connection: close")
    .appendBodyResponse(bodyMessage)
    .send();
```

### void    sendHTMLBasicResponse(String message)
Builds a basic response (setting 200 as status code and sending a basic HTML template) with the message passed as parameter.

### void    sendNonImplementedMethodResponse(String message)
Builds a basic 405 error response.
* **message**: (Optional): Message to be sent in the body. The default message is:  "Error: The used HTTP method have not been implemented on this endpoint."

### void    sendNotFoundResponse(String message)
Builds a basic 404 error response.
* **message**: (Optional): Message to be sent in the body. The default message is:  "Error: Resource not found."

### void    enableLogs()
Allow inner logs to be show in the serial monitor

###void    disableLogs()
Disable inner logs to be show in the serial monitor. This is the default behavior


changelog.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0] - 2016-03-07
### Changed
- Request class becomes ClientHandler

### Added
- ServerHandler class API, allowing to build a server to listen requests, build and send responses.
- Basic ServerHandler example
- ServoMotor management via a REST API build using ServerHandler

## [0.1.1] - 2016-02-28
### Changed
- init signature, from now it returns a bool.

## 0.1.0 - 2016-02-28
### Added
- README
- GET basic example.
- Basics of the library API
	* Constructor
	* init
	* Methods to handle request = initRequest, addHeader, send
	* Return statusCode and body (const char* format)

[1.0.0]: https://github.com/jomaora/schapiuino/compare/v0.1.1...v1.0.0
[0.1.1]: https://github.com/jomaora/schapiuino/compare/v0.1.0...v0.1.1



Request.h
/*
  Request.h - Library for create Http Requests.
  Created by Joan Ortega, jomaora@gmail.com Feb. 15, 2016.
  Released into the public domain.
*/

#ifndef Request_h
#define Request_h

#include "Arduino.h"
#include <Ethernet.h>

class Request
{
    public:
        Request(EthernetClient *client, byte mac[], int ipRanges[]){
        	_client = client;
          for (int i = 0; i < 6; i++) {
            _mac[i] = mac[i];  
          }
          _ip = IPAddress(ipRanges[0], ipRanges[1], ipRanges[2], ipRanges[3]);
          _requestReady = false;
        }

        bool    init();
        bool    initRequest(String httpMethod, char *server, int severPort, String path);
        void    addHeader(String header);
        const char* send();
        
        void    setPath(String path){ _path = path; }
        String  getPath() { return _path; }
        int     getResponseStatusCode();
        const char* getResponseBody();

        void    enableLogs() { _activeLogs = true; }
        void    disableLogs() { _activeLogs = false; }

    private:
      void readingResponse(char n);
      void buildResponse();
      void resetRequest();

    	EthernetClient *_client;
      byte _mac[6];
      IPAddress _ip;

    	char*  _server;
      int    _serverPort;
      String _path;
      String _headers[];

      bool _requestReady = false;
      bool _receivedResponse = false;

    	bool _requestSent;
    	bool _readingStatus = false;
    	bool _readingheaders = false;
    	bool _readingBody = false;

    	/*Variables to store response from API*/
    	String _status = "";
  		String _body = "";
      String _responseHeaders = "";

  		/*Counter on \n and \r characters sent by the API in the response, to identify which element have been read*/
		  int _returnCharCount = 0;

      bool  _activeLogs = false;
};
#endif

Request.cpp
/*
  Request.cpp - Library for create Http Requests.
  Created by Joan Ortega, jomaora@gmail.com Feb. 15, 2016.
  Released into the public domain.
*/

#include "Arduino.h"
#include <Ethernet.h>
#include "Request.h"

bool Request::init() {
  _requestReady = false;
  String result = "";
  Serial.begin(9600);
   while (!Serial) {
    ; // wait for serial port to connect. Needed for Leonardo only
  }

  int beginOK = Ethernet.begin(_mac);
  if (beginOK == 0) {
    result = "Failed to configure Ethernet using DHCP";
    Serial.println(result);
    Ethernet.begin(_mac, _ip);
  }
  delay(1000);
  result = "Ethernet Shield connected";
  Serial.println(result);
  return (beginOK != 0);
}

bool Request::initRequest(String httpMethod, char *server, int serverPort, String path) {
  _server = server;
  _path = path;
  resetRequest();

  int connected = 0;
  while (!connected) {
    Serial.println("Trying connection");
    if (_client->connect(_server, serverPort)) {
      Serial.println("connected");
      connected = 1;
    } 
    else {
      Serial.println("connection failed");
      _requestReady = false;
      return _requestReady;
    }
  }

  String requestLine = httpMethod + " ";
  requestLine.concat(path);
  requestLine.concat(" HTTP/1.1");
  String hostLine = "Host: " + String(server);
  _client->println(requestLine);
  _client->println(hostLine);

  _requestReady = true;
  return _requestReady;
}

void Request::addHeader(String header) {
  if (_requestReady) {
    _client->println(header);  
  }
}

const char* Request::send() {
  if (_requestReady) {
    _client->println("Connection: close");
    _client->println();

    buildResponse();

    if (_activeLogs) {
      Serial.println("disconnecting.");
      Serial.println("");
      Serial.println("status");
      Serial.println(_status);
      Serial.println("");
      Serial.print("headers");
      Serial.print(_responseHeaders);
      Serial.println("");
      Serial.print("body");
      Serial.println(_body);
    }

    _client->stop();

    return _body.c_str();
  }
  else {
    return "Request is not ready";
  }
}

void Request::buildResponse() {
  // values needed before reading response
  _returnCharCount = 0; _readingStatus = true;  _readingheaders = false;  _readingBody = false;
  while(!_receivedResponse) {
    if (_client->available()) {
      char c = _client->read();
      readingResponse(c);
    }
    if (!_client->available() && !_client->connected()) {
      _receivedResponse = true;
    }
  }
}

void Request::readingResponse(char c) {
  if (c == '\n' || c == '\r') {
    _returnCharCount++;
  } else {
    _returnCharCount = 0;
  }

  if (_returnCharCount == 2) {
    _readingStatus = false; _readingheaders = true;   _readingBody = false;
  }
  if (_returnCharCount == 4) {
    _readingStatus = false; _readingheaders = false;  _readingBody = true;
  }
  
  if (_readingStatus) {
    _status.concat(c);
  }
  if (_readingBody && c != '\n') {
    _body.concat(c);
  } 
  if (_readingheaders) {
    _responseHeaders.concat(c);  
  }
}

int Request::getResponseStatusCode() {
  if (_receivedResponse) {
    int blankSpaceIndex = _status.indexOf(" ") + 1;
    String statusCode = _status.substring(blankSpaceIndex, blankSpaceIndex + 3);
    return statusCode.toInt();  
  } else {
    return 0;
  }
}

const char* Request::getResponseBody() {
  if (_receivedResponse) {
    return _body.c_str();  
  } else {
    return "";
  }
}

void Request::resetRequest() {
  _receivedResponse = false;  // setting the flag that indicates the response recpetion to false;
  _body = "";
  _status = "";
  _responseHeaders = "";
}

ClientHandler.h
/*
  ClientHandler.h - Library for create and send Http Requests.
  Created by Joan Ortega, jomaora@gmail.com Feb. 15, 2016.
  Released into the public domain.
*/

#ifndef ClientHandler_h
#define ClientHandler_h

#include "Arduino.h"
#include <Ethernet.h>

class ClientHandler
{
    public:
        ClientHandler(EthernetClient *client, byte mac[], int ipRanges[]){
            _client = client;
            for (int i = 0; i < 6; i++) {
                _mac[i] = mac[i];
            }
            _ip = IPAddress(ipRanges[0], ipRanges[1], ipRanges[2], ipRanges[3]);
            _requestReady = false;
        }

        bool    init();
        bool    initClientHandler(String httpMethod, char *server, int severPort, String path);
        void    addHeader(String header);
        const char* send();
        
        void    setPath(String path) { _path = path; }
        String  getPath() { return _path; }
        int     getResponseStatusCode();
        const char* getResponseBody();

        void    enableLogs() { _activeLogs = true; }
        void    disableLogs() { _activeLogs = false; }

    private:
        void readingResponse(char n);
        void buildResponse();
        void resetRequest();

        EthernetClient *_client;
        byte _mac[6];
        IPAddress _ip;

        char*  _server;
        int    _serverPort;
        String _path;
        String _headers[];

        bool _requestReady = false;
        bool _receivedResponse = false;

        bool _requestSent;
        bool _readingStatus = false;
        bool _readingheaders = false;
        bool _readingBody = false;

        /*Variables to store response from API*/
        String _status = "";
        String _body = "";
        String _responseHeaders = "";

        /*Counter on \n and \r characters sent by the API in the response, to identify which element have been read*/
        int _returnCharCount = 0;

        bool  _activeLogs = false;
};
#endif

ServerHandler.h
/*
  ServerHandler.h - Library for receive Http Requests and send Http responses.
  Created by Joan Ortega, jomaora@gmail.com Feb. 29, 2016.
  Released into the public domain.
*/

#ifndef ServerHandler_h
#define ServerHandler_h

#include "Arduino.h"
#include <Ethernet.h>

class ServerHandler
{
    public:
        ServerHandler(EthernetServer *server, byte mac[], int ipRanges[]) {
            _server = server;
            for (int i = 0; i < 6; i++) {
                _mac[i] = mac[i];
            }
            _ip = IPAddress(ipRanges[0], ipRanges[1], ipRanges[2], ipRanges[3]);
            _requestReceived = false;
        }

        void    init();
        bool    listenRequest();

        ServerHandler & buildResponse(int statusCode, String message = "");
        ServerHandler & appendHeaderResponse(String header);
        ServerHandler & appendBodyResponse(String body);

        void    send();
        void    sendHTMLBasicResponse(String message);
        void    sendNonImplementedMethodResponse(String message = "Error: The used HTTP method have not been implemented on this endpoint.");
        void    sendNotFoundResponse(String message = "Error: Resource not found.");

        String      getHttpMethod();
        String      getRequestPath();
        const char* getRequestHeaders();
        const char* getRequestBody();

        void    enableLogs() { _activeLogs = true; }
        void    disableLogs() { _activeLogs = false; }

    private:
        void    readRequest();
        void    resetRequest();
        void    parseStatusCode(int status, String message = "");

        EthernetServer *_server;
        EthernetClient _client;
        byte _mac[6];
        IPAddress _ip;

        bool _requestReceived = false;
        int  _readingStatus = 0;

        /*Variables to store request from Client*/
        String _requestLine = "";
        String _requestBody = "";
        String _requestHeaders = "";

        /*Counter on \n and \r characters sent by the API in the response, to identify which element have been read*/
        int  _returnCharCount = 0;
        bool _activeLogs = false;
};
 #endif

ClientHandler.cpp
/*
  ClientHandler.cpp - Library for create Http Requests.
  Created by Joan Ortega, jomaora@gmail.com Feb. 15, 2016.
  Released into the public domain.
*/

#include "Arduino.h"
#include <Ethernet.h>
#include "ClientHandler.h"

bool ClientHandler::init() {
  _requestReady = false;
  String result = "";
   while (!Serial) {
    ; // wait for serial port to connect. Needed for Leonardo only
  }

  int beginOK = Ethernet.begin(_mac);
  if (beginOK == 0) {
    result = "Failed to configure Ethernet using DHCP";
    Serial.println(result);
    Ethernet.begin(_mac, _ip);
  }
  delay(1000);
  result = "Ethernet Shield connected";
  Serial.println(result);
  return (beginOK != 0);
}

bool ClientHandler::initClientHandler(String httpMethod, char *server, int serverPort, String path) {
  _server = server;
  _path = path;
  resetRequest();

  int connected = 0;
  while (!connected) {
    Serial.println("Trying connection");
    if (_client->connect(_server, serverPort)) {
      Serial.println("connected");
      connected = 1;
    } 
    else {
      Serial.println("connection failed");
      _requestReady = false;
      return _requestReady;
    }
  }

  String requestLine = httpMethod + " ";
  requestLine.concat(path);
  requestLine.concat(" HTTP/1.1");
  String hostLine = "Host: " + String(server);
  _client->println(requestLine);
  _client->println(hostLine);

  _requestReady = true;
  return _requestReady;
}

void ClientHandler::addHeader(String header) {
  if (_requestReady) {
    _client->println(header);  
  }
}

const char* ClientHandler::send() {
  if (_requestReady) {
    _client->println("Connection: close");
    _client->println();

    buildResponse();

    if (_activeLogs) {
      Serial.println("disconnecting.");
      Serial.println("");
      Serial.println("status");
      Serial.println(_status);
      Serial.println("");
      Serial.print("headers");
      Serial.print(_responseHeaders);
      Serial.println("");
      Serial.print("body");
      Serial.println(_body);
    }

    _client->stop();

    return _body.c_str();
  }
  else {
    return "Request is not ready to be sent";
  }
}

void ClientHandler::buildResponse() {
  // values needed before reading response
  _returnCharCount = 0; _readingStatus = true;  _readingheaders = false;  _readingBody = false;
  while(!_receivedResponse) {
    if (_client->available()) {
      char c = _client->read();
      readingResponse(c);
    }
    if (!_client->available() && !_client->connected()) {
      _receivedResponse = true;
    }
  }
}

void ClientHandler::readingResponse(char c) {
  if (c == '\n' || c == '\r') {
    _returnCharCount++;
  } else {
    _returnCharCount = 0;
  }

  if (_returnCharCount == 2) {
    _readingStatus = false; _readingheaders = true;   _readingBody = false;
  }
  if (_returnCharCount == 4) {
    _readingStatus = false; _readingheaders = false;  _readingBody = true;
  }
  
  if (_readingStatus) {
    _status.concat(c);
  }
  if (_readingBody && c != '\n') {
    _body.concat(c);
  } 
  if (_readingheaders) {
    _responseHeaders.concat(c);  
  }
}

int ClientHandler::getResponseStatusCode() {
  if (_receivedResponse) {
    int blankSpaceIndex = _status.indexOf(" ") + 1;
    String statusCode = _status.substring(blankSpaceIndex, blankSpaceIndex + 3);
    return statusCode.toInt();  
  } else {
    return 0;
  }
}

const char* ClientHandler::getResponseBody() {
  if (_receivedResponse) {
    return _body.c_str();  
  } else {
    return "";
  }
}

void ClientHandler::resetRequest() {
  _receivedResponse = false;  // setting the flag that indicates the response recpetion to false;
  _body = "";
  _status = "";
  _responseHeaders = "";
}

ServerHandler.cpp
/*
  ServerHandler.cpp - Library for receive Http Requests and send Http responses.
  Created by Joan Ortega, jomaora@gmail.com Feb. 29, 2016.
  Released into the public domain.
*/

#include "Arduino.h"
#include <Ethernet.h>
#include "ServerHandler.h"

void ServerHandler::init() {
    while (!Serial) {
        ; // wait for serial port to connect. Needed for Leonardo only
    }

    Ethernet.begin(_mac, _ip);
    _server->begin();
    if (_activeLogs) {
        Serial.print("server is at ");
        Serial.println(Ethernet.localIP());
    }
}

bool ServerHandler::listenRequest() {
    _client = _server->available();
    if (!_client) {
        return false;
    } else {
        if (_activeLogs) {
            Serial.println("incomming request");
        }
        resetRequest();
        readRequest();
        return true;
    }
}

void ServerHandler::sendHTMLBasicResponse(String message) {
    if (_requestReceived) {
        _client.println("HTTP/1.1 200 OK");
        _client.println("Content-Type: text/html");
        _client.println("Connection: close");
        _client.println();
        _client.println("<!DOCTYPE HTML>");
        _client.println("<html>");
        _client.println(message);
        _client.println("</html>");
        delay(1);
        _client.stop();
        if (_activeLogs) {
            Serial.println("client disconnected");
        }
    }
}

ServerHandler & ServerHandler::buildResponse(int statusCode, String message) {
    if (_requestReceived) {
        parseStatusCode(statusCode, message);
    }
    return *this;
}

ServerHandler & ServerHandler::appendHeaderResponse(String header) {
    if (_requestReceived) {
        _client.println(header);
    }
    return *this;
}

ServerHandler & ServerHandler::appendBodyResponse(String body) {
    if (_requestReceived) {
        _client.println();
        _client.println(body);
    }
    return *this;
}

void ServerHandler::send() {
    if (_requestReceived) {
        delay(1);
        _client.stop();
        if (_activeLogs) {
            Serial.println("client disconnected");
        }
    }
}

String ServerHandler::getHttpMethod() {
    if (_requestReceived) {
        int blankSpaceIndex = _requestLine.indexOf(" ");
        String httpMethod = _requestLine.substring(0, blankSpaceIndex);
        return _requestLine.substring(0, blankSpaceIndex);
    } else {
        return "";
    }
}

String ServerHandler::getRequestPath() {
    if (_requestReceived) {
        int firstBlankSpaceIndex = _requestLine.indexOf(" ") + 1;
        int secondBlankSpaceIndex = _requestLine.indexOf(" ", firstBlankSpaceIndex);
        String path = _requestLine.substring(firstBlankSpaceIndex, secondBlankSpaceIndex);
        return _requestLine.substring(firstBlankSpaceIndex, secondBlankSpaceIndex);
    } else {
        return "";
    }
}

const char* ServerHandler::getRequestHeaders() {
    return _requestHeaders.c_str();
}

const char* ServerHandler::getRequestBody() {
    return _requestBody.c_str();
}

void ServerHandler::readRequest() {
    boolean currentLineIsBlank = true;
    int countCLRF = 0;
    while (_client.connected() && !_requestReceived) {
        if (_client.available()) {
            char c = _client.read();
            if (c == '\n' || c == '\r') {
                _returnCharCount++;
            } else {
                _returnCharCount = 0;
            }
            if (_returnCharCount == 2) {
                _readingStatus = 1;
            }
            if (_returnCharCount == 4) {
                _readingStatus = 2;
            }

            switch (_readingStatus) {
                case 0:
                    _requestLine.concat(c);
                    break;
                case 1:
                    _requestHeaders.concat(c);
                    break;
                case 2:
                    if (c != '\n') {
                        _requestBody.concat(c);
                    }
                    break;
            }
        } else {
            _requestReceived = true;
        }
    }
}

void ServerHandler::resetRequest() {
    _readingStatus = 0;
    _returnCharCount = 0;
    _requestLine = "";
    _requestBody = "";
    _requestHeaders = "";
    _requestReceived = false;
}

void ServerHandler::parseStatusCode(int status, String message) {
    _client.print("HTTP/1.1 ");
    _client.print(status);
    switch (status) {
        case 200:
            _client.println("OK");
            break;
        case 201:
            _client.println("Accepted");
            break;
        case 204:
            _client.println("No Content");
            break;
        case 400:
            _client.println("Bad Request");
            break;
        case 401:
            _client.println("Unauthorize");
            break;
        case 403:
            _client.println("Forbidden");
            break;
        case 404:
            _client.println("Not Found");
            break;
        case 405:
            _client.println("Method Not Allowed");
            break;
        case 406:
            _client.println("Not Acceptable");
            break;
        case 500:
            _client.println("Internal server error");
            break;
        default:
            _client.println(message);
            break;
    }
}

void ServerHandler::sendNonImplementedMethodResponse(String message) {
    if (_requestReceived) {
        buildResponse(405)
            .appendHeaderResponse("Content-Type: text/plain")
            .appendHeaderResponse("Connection: close")
            .appendBodyResponse(message)
            .send();
        delay(1);
        _client.stop();
    }
}

void ServerHandler::sendNotFoundResponse(String message) {
    if (_requestReceived) {
        buildResponse(405)
            .appendHeaderResponse("Content-Type: text/plain")
            .appendHeaderResponse("Connection: close")
            .appendBodyResponse(message)
            .send();
        delay(1);
        _client.stop();
    }
}